### PR TITLE
Add check to make sure only strings are assigned to __name__

### LINKED ISSUE #583: [easy] can assign only strings to __name__
Originally reported by: **Claudiu Popa (BitBucket: [PCManticore](http://bitbucket.org/PCManticore), GitHub: @PCManticore)**

---

```
#!python

def test(): pass
test.__name__ = 1
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/583

### PR DESCRIPTION
**Pull request by ninezerozeronine**
## Steps

- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description

This adds a new check to make sure that only strings are assigned to `__name__` attributes.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

## Related Issue

Closes #583

### CODE REVIEW

**Comment by ninezerozeronine** [comment]
Putting this in as a draft to check I'm going in the right direction.

I'd like to:

- Add inference to detect things like:
```
def return_string():
    return "foo"

def my_func():
    pass

my_func.__name__ = return_string()
```

- Investigate how using inference relates to confidence levels and add this to the check output if necessary.
- Improve check output to include more details of why the check failed.

File: pylint/checkers/typecheck.py [hunk 1]
```diff
@@ -1035,6 +1053,25 @@ def visit_assign(self, node):
             else:
                 self.add_message("assignment-from-none", node=node)
 
+    def _check_dundername_is_string(self, node):
+        """
+        Check a string is assigned to self.__name__
+        """
+
+        # Check the left hand side of the assignment is <something>.__name__
+        lhs = list(node.get_children())[0]
```
Comment (PCManticore) [inline pylint/checkers/typecheck.py]: We could inspect `node.targets` instead to get the LHS of the assignment.

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -1035,6 +1053,25 @@ def visit_assign(self, node):
...
+        if not isinstance(lhs, astroid.node_classes.AssignAttr):
+            return
+        if not lhs.attrname == "__name__":
+            return
+
+        # If the right hand side is not a string
+        rhs = list(node.get_children())[1]
```
Comment (PCManticore) [inline pylint/checkers/typecheck.py]: That will be `node.value` instead.

File: pylint/checkers/typecheck.py [hunk 3]
```diff
@@ -1035,6 +1053,25 @@ def visit_assign(self, node):
...
+        if not (isinstance(rhs, astroid.node_classes.Const)
```
Comment (PCManticore) [inline pylint/checkers/typecheck.py]: Unfortunately it's not quite that simple, we need to infer the value on the right hand side to figure out if it is a string or not. To speed it up, we can still use the `Const` guard, so something along these lines should do the trick:
```
if isinstance(rhs, astroid.Const) and isinstance(rhs.value, str):
     return
inferred = utils.safe_infer(rhs)
if not inferred:
    return
if not (isinstance(inferred, astroid.Const) and isinstance(inferred.value, str)):
    # emit the message
```

File: tests/functional/n/non_str_assignment_to_dunder_name.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+class ExampleClass():
+    pass
+
+
+def example_function():
+    pass
+
+
+ExampleClass.__name__ = 1  # [non-str-assignment-to-dunder-name]
```
Comment (PCManticore) [inline tests/functional/n/non_str_assignment_to_dunder_name.py]: Let's add a couple of additional test cases:

* One with a function call on the RHS returning a string (should not emit)
* One with a function call on the RHS returning an integer (should emit)
* One with a functional on the RHS returning multiple types (should not emit)
* Do `from unknown import Unknown` and place `Unknown` on the RHS (should not emit)

**Comment by ninezerozeronine** [comment]
Thank you for the advice and examples Claudiu! I'm hoping to take a stab at this this weekend.

File: pylint/checkers/typecheck.py (hunk 3 above)
Comment (ninezerozeronine) [reply pylint/checkers/typecheck.py]: Thank you! Should be fixed in https://github.com/PyCQA/pylint/pull/3271/commits/abbcafab1f23b72cfc37af10861d3ded58dd7a6d

File: pylint/checkers/typecheck.py (hunk 1 above)
Comment (ninezerozeronine) [reply pylint/checkers/typecheck.py]: Thank you! Should be fixed in https://github.com/PyCQA/pylint/pull/3271/commits/abbcafab1f23b72cfc37af10861d3ded58dd7a6d

File: pylint/checkers/typecheck.py (hunk 2 above)
Comment (ninezerozeronine) [reply pylint/checkers/typecheck.py]: Thanks! Should be fixed in https://github.com/PyCQA/pylint/pull/3271/commits/abbcafab1f23b72cfc37af10861d3ded58dd7a6d

File: tests/functional/n/non_str_assignment_to_dunder_name.py (hunk 4 above)
Comment (ninezerozeronine) [reply tests/functional/n/non_str_assignment_to_dunder_name.py]: Thank you!

I wasn't quite sure what 

> One with a functional on the RHS returning multiple types (should not emit)

meant, but I added a return of a tuple, and a return that coule be multiple types based on runtime conditions. Hope that covers it!

Changes are in https://github.com/PyCQA/pylint/pull/3271/commits/b4f795582bf924384517faa46aa89dc62fc5b086

**Comment by PCManticore** [comment]
Thanks a lot for the patch @ninezerozeronine !

**Pull request merged**